### PR TITLE
Home task 04 completed

### PR DESCRIPTION
**Pull request by jhonfmg7**
Home task 04 completed,
Project created without "create-react-app"
using typescript, CSS modules
creating add, edit, delete modals and sorting functionality

Captura de pantalla 2022-11-03 a la(s) 9 16 12 a m

### CODE REVIEW

File: src/components/header/index.jsx [hunk 1]
```diff
@@ -17,12 +19,15 @@ const Header = () => {
...
+        <header id="header" className={ styles.hero } style={{ backgroundImage: 'url(/img/hero.svg)' }} data-testid="header">
+            { isOpenModal && (
+                <AddOrEditModal title="add" setIsOpen={ setIsOpenModal } />
+            ) }
             <div className={ styles.container }>
                 <a href="#" className={ styles.noTextDecoration }>
                     <Logo />
                 </a>
-                <button className={ styles.mainButton }><span>+ </span>ADD MOVIE</button>
+                <button className={ styles.mainButton } onClick={ () => setIsOpenModal(!isOpenModal) }><span>+ </span>ADD MOVIE</button>
```
Comment (mmarkelov) [inline src/components/header/index.jsx]: Here you are related on the previous state - it's better to use function:
```js
setIsOpenModal(prev => !prev)
```

File: src/components/main/CardMovie.jsx [hunk 2]
```diff
@@ -1,22 +1,64 @@
...
+                        rating: movie.rating,
+                        genre: movie.genre,
+                        runtime: movie.runtime,
+                        overview: movie.overview
+                    }} 
+                />
+            ) }
+            <article className={ styles.movieItem }>
+                <div className={ styles.menuContainer }>
+                    <div className={ styles.menuButton } onClick={ () => setIsOpenMenu(!isOpenMenu) }>...</div>
```
Comment (mmarkelov) [inline src/components/main/CardMovie.jsx]:
```suggestion
                    <div className={ styles.menuButton } onClick={ () => setIsOpenMenu(prevState => !prevState) }>...</div>
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    date: String,
+    url: String,
+    rating: Number,
+    genre: String,
+    runtime: Number,
+    overview: string
+}
+
+interface Props {
+    movie?: any
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: Try to avoid `any`

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
+import styles from '../../../css-modules/modal.module.css';
+import stylesHeader from '../../../css-modules/header.module.css';
+
+// Components
+import Input from './Input';
+
+interface InitialState {
+    title: String,
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: shouldn't it be?
```suggestion
    title: string,
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    date: String,
+    url: String,
+    rating: Number,
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: same here
```suggestion
    rating: number,
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        url: "",
+        rating: 0,
+        genre: "",
+        runtime: 0,
+        overview: ""
+
+    }
+    const [ info, setInfo ] = React.useState(initialState);
+
+    const handleChange = (e:React.ChangeEvent<HTMLInputElement>) => {
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: I guess that you can use one function for input and textarea

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    const handleReset = () => setInfo(initialState);
+
+    React.useEffect(() => {
+        if (movie) {
+            setInfo(movie)
+        } else {
+            setInfo(initialState);
+        }
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]:
```suggestion
            setInfo(movie ?? initialState)
```

File: src/components/modals/Modal.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import styles from '../../css-modules/modal.module.css';
+
+interface Props {
+    children: React.ReactNode,
+    title: String,
+    setIsOpen: Function, 
+    setIsOpenMenu: Function
```
Comment (mmarkelov) [inline src/components/modals/Modal.tsx]: shouldn't it be optional?

**Comment by mmarkelov** [comment]
Some general points:
- You use the mix of JS and TS, please use only TS for your code
- It's better to use more specific types instead of `Function`

File: src/components/header/index.jsx (hunk 1 above)
Comment (jhonfmg7) [reply src/components/header/index.jsx]: resolved

File: src/components/main/CardMovie.jsx (hunk 2 above)
Comment (jhonfmg7) [reply src/components/main/CardMovie.jsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 3 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 4 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 5 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 6 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 7 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/Modal.tsx (hunk 8 above)
Comment (jhonfmg7) [reply src/components/modals/Modal.tsx]: resolved

File: src/components/main/CardMovie.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import MovieMenu from './MovieMenu';
+import DeleteModal from '../modals/DeleteModal';
+import AddOrEditModal from '../modals/AddOrEditModal';
+import MoreInfoMovie from '../modals/MoreInfoMovie';
+
+interface Movie {
+    image: string,
+    title: string,
+    rating: number,
+    genres: Array<string>,
```
Comment (mmarkelov) [inline src/components/main/CardMovie.tsx]: Alternatively it can be `string[]`

File: src/components/main/MovieMenu/index.tsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import * as React from 'react';
+import styles from '../../../css-modules/main.module.css';
+import stylesModal from '../../../css-modules/modal.module.css';
+
+interface Props {
+    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>,
+    setIsOpenDeleteModal: React.Dispatch<React.SetStateAction<boolean>>,
+    setIsOpenEditModal: React.Dispatch<React.SetStateAction<boolean>>
+}
```
Comment (mmarkelov) [inline src/components/main/MovieMenu/index.tsx]: usually it's better to use this:
```suggestion
interface Props {
    setIsOpen: (newState:boolean) => void,
    setIsOpenDeleteModal: (newState:boolean) => void,
    setIsOpenEditModal: (newState:boolean) => void
}
```
cause child component don't really need to know what the passed function do in parent component

File: src/components/main/navbar/index.tsx [hunk 11]
```diff
@@ -1,15 +1,24 @@
...
 import camelCase from '../../../utils/camelCase';
 
 const CATEGORIES = [ 'all', 'action', 'documentary', 'comedy', 'horror', 'crime' ];
 
-const NavBar = ({ typeSelected, setTypeSelected, sortBySelected, setSortBySelected }) => {
+interface Props {
+    typeSelected: string, 
+    sortBySelected: string,
+    setTypeSelected: React.Dispatch<React.SetStateAction<string>>,
+    setSortBySelected: React.Dispatch<React.SetStateAction<string>>
```
Comment (mmarkelov) [inline src/components/main/navbar/index.tsx]:
```suggestion
    setTypeSelected: (newType: string) => void,
    setSortBySelected: (newValue: string) => void,
```

File: src/components/modals/AddOrEditModal.tsx [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    genre: string | Array<string>,
+    runtime: number,
+    overview: string
+}
+
+interface Props {
+    title: string,
+    movieInfo?: Movie,
+    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>,
+    setIsOpenMenu?: React.Dispatch<React.SetStateAction<boolean>>
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditModal.tsx]:
```suggestion
    setIsOpen: (newState: boolean) => void,
    setIsOpenMenu?: (newState: boolean) => void,
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
+import * as React from 'react';
+import styles from '../../../css-modules/modal.module.css';
+import stylesHeader from '../../../css-modules/header.module.css';
+
+// Components
+import Input from './Input';
+
+interface InitialState {
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: it's just the same type as `Movie` in [AddOrEditModal.tsx](https://github.com/jhonfmg7/home-task-02/pull/3/files#diff-c97e9f2916d9231ecc96b081e08f7c41b650e554a5052cfdad3e87c2767c5330). I guess that you can create a `types` directory to store and reuse some comment types between components

File: src/components/modals/DeleteModal.tsx [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+import styles from '../../css-modules/modal.module.css';
+import stylesHeader from '../../css-modules/header.module.css';
+
+// Components
+import Modal from './Modal';
+
+interface Props {
+    title: string,
+    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>,
+    setIsOpenMenu?: React.Dispatch<React.SetStateAction<boolean>>
```
Comment (mmarkelov) [inline src/components/modals/DeleteModal.tsx]:
```suggestion
    setIsOpen: (newState: boolean) => void,
    setIsOpenMenu?: (newState: boolean) => void,
```

File: src/components/modals/Modal.tsx [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import styles from '../../css-modules/modal.module.css';
+
+interface Props {
+    children: React.ReactNode,
+    title: string,
+    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>, 
+    setIsOpenMenu: React.Dispatch<React.SetStateAction<boolean>>
```
Comment (mmarkelov) [inline src/components/modals/Modal.tsx]:
```suggestion
    setIsOpen: (newState: boolean) => void, 
    setIsOpenMenu: (newState: boolean) => void, 
```

Comment (jhonfmg7) [reply src/components/modals/Modal.tsx]: fixed

File: src/components/modals/DeleteModal.tsx (hunk 14 above)
Comment (jhonfmg7) [reply src/components/modals/DeleteModal.tsx]: fixed

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 13 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: Thank you, I resolved this by creating a new types folder to add interfaces

File: src/components/modals/AddOrEditModal.tsx (hunk 12 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditModal.tsx]: fixed

File: src/components/main/navbar/index.tsx (hunk 11 above)
Comment (jhonfmg7) [reply src/components/main/navbar/index.tsx]: fixed

File: src/components/main/MovieMenu/index.tsx (hunk 10 above)
Comment (jhonfmg7) [reply src/components/main/MovieMenu/index.tsx]: fixed

File: src/components/main/CardMovie.tsx (hunk 9 above)
Comment (jhonfmg7) [reply src/components/main/CardMovie.tsx]: fixed

**Pull request merged**